### PR TITLE
Hack Week: Store Order Notifications

### PR DESCRIPTION
**Pull request by Ecarrion**
# Note

Apologies for the Big PR, as I did not know if this project was going to be shipped successfully, I was not able to break it efficiently beforehand 🙏 

# Why

For my Hack Week project, I decided to explore providing a richer UI for order notifications. 

# How

- Adds a Notification Content Extension Target. This is where the notification UI runs as it is not part of the main app.

- Read credentials from the keychain: This is needed to fire requests. We only care about wpcom credentials as push notifications only work when jetpack is active and connected.

- When a user long presses on an order notification, the extension:
    - Loads the notification object from the server
    - Loads the order object from the server
    - Formats and renders the richer UI.
    
# How to review

The code lives in three different files:
 - `OrderNotificationViewController`
 - `OrderNotificationViewModel`
 - `OrderNotificationView`
 
The other files are just assets, configurations, and a required storyboard as an entry point.

# How to test

- Download the notification extensions development profile from our development-profile bundle.
- Run the app into the device, and make sure notifications are enabled
- Close the app and create an order via web to the store.
- Wait for the push notification to arrive and long press on it.
- See the custom UI for the order notification.

# Screenshots
Light | Dark 
---- | ----
light| dark

    
---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by Ecarrion** [comment]
Hi @spencertransier! 

The installable/prototype build is failing in the gym step. It must be due to the new Notifications extensions but I don't think I have the necessary resources to fix the issue.

Do you know if something is missing? I'm happy to help with anything 😃

**Comment by spencertransier** [comment]
@Ecarrion Thanks for letting me know! I will take a look at the failing build and see what's happening.

**Comment by spencertransier** [comment]
@Ecarrion It ended up being an issue with the alpha bundle identifier and assigning the correct provisioning profile for the `Release-Alpha` setting in Xcode. Looks like everything is working now 👍

**Comment by Ecarrion** [comment]
> @Ecarrion It ended up being an issue with the alpha bundle identifier and assigning the correct provisioning profile for the `Release-Alpha` setting in Xcode. Looks like everything is working now 👍

Thank you very much @spencertransier!

**Review by jaclync — COMMENTED** [review]
Looks much nicer! Just some high-level questions before testing and reviewing the code deeply:

- Does this change affect the new review notifications?
- Does this change affect other scenarios in `PushNotificationsManager` - handling local/remote notifications in the foreground/background?
- Before this change, does the app currently make an API request to fetch the order for each push notification? I'm a bit concerned by stores with high order volume that can cause lots of API requests and potentially impact the site performance. --> an alternative is to add the extra order fields (like `datePaid/dateCreated/total/currencySymbol/paymentMethodTitle/shippingLables`) or even just return the strings in the server-side. However, I'm not sure if this conflicts with our effort to support push notifications without Jetpack (with WPCOM backend)

On the side:
- Do we want to wait for design confirmation?

**Comment by Ecarrion** [comment]
Hey @jaclync 


> Does this change affect the new review notifications?

No, it only affects the [store order](https://github.com/woocommerce/woocommerce-ios/pull/10658/files#diff-7d4b6936c49687a411e08f4062456f33a42c85c9feda067c9da3a1cc00b40510R10) notifications.


> Does this change affect other scenarios in PushNotificationsManager - handling local/remote notifications in the foreground/background?

Nope.

> Before this change, does the app currently make an API request to fetch the order for each push notification? I'm a bit concerned by stores with high order volume that can cause lots of API requests and potentially impact the site performance. 

No, all network requests happen on demand, when the user long presses on a store order notification. So getting a high number of notifications won't automatically fire any network request.


> Do we want to wait for design confirmation?

This design was provided by Joe, so we are good to go!

File: WooCommerce/NotificationExtension/Info.plist [hunk 1]
```diff
@@ -0,0 +1,22 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
+<plist version="1.0">
+<dict>
+	<key>NSExtension</key>
+	<dict>
+		<key>NSExtensionAttributes</key>
+		<dict>
+			<key>UNNotificationExtensionCategory</key>
+			<string>store_order</string>
```
Comment (jaclync) [inline WooCommerce/NotificationExtension/Info.plist:10]: I'm curious about how only order notifications are routed to the notification extension, is this how notifications of type `store_order` come to this?

File: WooCommerce/NotificationExtension/OrderNotificationViewController.swift [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+            }
+        }
+    }
+
+    @MainActor
+    private func addOrderNotificationView(with content: OrderNotificationView.Content) {
+        let orderView = OrderNotificationView(content: content)
+        hostingView = UIHostingController(rootView: orderView)
+
+        self.view.addSubview(hostingView.view)
```
Comment (jaclync) [inline WooCommerce/NotificationExtension/OrderNotificationViewController.swift]: super nit:
```suggestion
        view.addSubview(hostingView.view)
```

File: WooCommerce/NotificationExtension/OrderNotificationViewController.swift [hunk 3]
```diff
@@ -0,0 +1,53 @@
+import UIKit
+import UserNotifications
+import UserNotificationsUI
+import SwiftUI
+
+class OrderNotificationViewController: UIViewController, UNNotificationContentExtension {
```
Comment (jaclync) [inline WooCommerce/NotificationExtension/OrderNotificationViewController.swift]: super nit: can this class be `final`?

File: WooCommerce/NotificationExtension/OrderNotificationViewController.swift [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+    @MainActor @IBOutlet var label: UILabel?
+    @MainActor @IBOutlet var loadingIndicator: UIActivityIndicatorView?
```
Comment (jaclync) [inline WooCommerce/NotificationExtension/OrderNotificationViewController.swift]: super nit: can these be private?

File: WooCommerce/NotificationExtension/OrderNotificationViewController.swift [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+                let note = try await viewModel.loadNotification(notification)
+                let order = try await viewModel.loadOrder(for: note)
+                let content = viewModel.formatContent(note: note, order: order)
+                addOrderNotificationView(with: content)
+                loadingIndicator?.isHidden = true
+
+            } catch {
+
+                loadingIndicator?.isHidden = true
+                self.label?.text = AppLocalizedString("Unable to load notification",
```
Comment (jaclync) [inline WooCommerce/NotificationExtension/OrderNotificationViewController.swift]: super nit: is `self` needed?

File: WooCommerce/NotificationExtension/OrderNotificationViewModel.swift [hunk 6]
```diff
@@ -0,0 +1,154 @@
+import Foundation
+import KeychainAccess
+import Networking
+
+/// View Model for the `OrderNotificationViewController`type.
+///
+final class OrderNotificationViewModel {
```
Comment (jaclync) [inline WooCommerce/NotificationExtension/OrderNotificationViewModel.swift:7]: non-blocking: would be nice to have some test cases!

**Review by jaclync — APPROVED** [review]
Works nicely! 🎉 

> No, all network requests happen on demand, when the user long presses on a store order notification. So getting a high number of notifications won't automatically fire any network request.

For future considerations: would be great if the order/note can be cached, so that long pressing the same notification multiple times doesn't make the same number of API requests. But it's tricky when the notification content is an app extension.

File: WooCommerce/NotificationExtension/Info.plist (hunk 1 above)
Comment (Ecarrion) [reply WooCommerce/NotificationExtension/Info.plist:10]: All push notifications that we send come with a `type` key. In this case `"type": "store_order"`
By setting that category in the extension info plist in the `UNNotificationExtensionCategory` we guarantee that only those type of notifications are routed to the notification extension.

Does that make sense?

File: WooCommerce/NotificationExtension/OrderNotificationViewModel.swift (hunk 6 above)
Comment (Ecarrion) [reply WooCommerce/NotificationExtension/OrderNotificationViewModel.swift:7]: Yeah, I wanted those but ran out of time in the HACK week 😇 
I have added an issue to the eagle backlog https://github.com/woocommerce/woocommerce-ios/issues/10683

File: WooCommerce/NotificationExtension/Info.plist (hunk 1 above)
Comment (jaclync) [reply WooCommerce/NotificationExtension/Info.plist:10]: Yup makes sense, thanks for the explanation!

**Pull request merged**